adds support for using connection=netconf

privateip opened the pull request:
This change updates the module to provide support for using
connection=netconf instead of connection=local.  If connection=netconf
is used, then the various connection arguments will be silently ignored.

Commit:
adds support for using connection=netconf

This change updates the module to provide support for using
connection=netconf instead of connection=local.  If connection=netconf
is used, then the various connection arguments will be silently ignored.

lpenz commented:
Looks great!
I'm not being able to test it, though. After changing some things to move the code forward, I end up getting something like
    AnsibleError: connection=netconf is not supported on <x>
...where <x> is whatever I set in ansible_network_os.

I understand that this is WIP, but let me know if I should not be seeing this.

privateip commented:
This PR is dependent on #33671

privateip commented:
@lpenz thats strange I cannot reproduce the error you are referencing.  There is a bug that is dependent on 33671 being merged but that is a different issue.  once 33671 is merged, could you please retest?

lpenz commented:
Will do, thanks!

privateip commented:
@lpenz do you have a chance to test this PR?

lpenz commented on lib/ansible/modules/network/netconf/netconf_config.py:
```diff
@@ -214,68 +244,82 @@ def main():
-    )
+    if not local_connection:
+        m = Connection(module._socket_path)
+        capabilities = module.from_json(m.get_capabilities())
```
I have been hitting an issue here, where `m` ends up being a module_utils.connection.Connection without a `get_capabilities` method. I'm using `ansible_network_os=default` to access an openyuma server. That works without using `ansible_connection=netconf`

privateip commented:
@lpenz that helps, thanks.  I see where the problem is and will have a fix shortly

Commit:
adds netconf plugin default

This adds a default implementation for netconf plugins if the network_os
is not specified.  The default plugin will implement only the standard
netconf rpcs

Commit:
fix up pep8 issues

privateip commented:
@lpenz I finally got around to putting that fix in.  This should now work without having to specify `ansible_network_os` variable.

lpenz commented:
I've tested it, and it's now working.
Thanks for doing this! It makes ansible look a lot more consistent when dealing with network devices.

lpenz commented:
I tried this out, and it is looking better, though I got another problem that will demand more debugging time.
(it's complaining that ncclient is not installed)
